silkrpc/core/rawdb/chain.cpp

```diff
@@ -98,6 +98,19 @@ asio::awaitable<intx::uint256> read_total_difficulty(const DatabaseReader& reade
     co_return total_difficulty;
 }
 
+asio::awaitable<silkworm::BlockWithHash> read_block_by_number_or_hash(const DatabaseReader& reader, const silkrpc::BlockNumberOrHash& bnoh) {
+    if (bnoh.is_number()) {
+        co_return co_await read_block_by_number(reader, bnoh.number());
+    } else if (bnoh.is_hash()) {
+        co_return co_await read_block_by_hash(reader, bnoh.hash());
+    } else if (bnoh.is_tag()) {
+        auto block_number = co_await get_latest_block_number(reader);
+        co_return co_await read_block_by_number(reader, block_number);
+    }
+
+    throw std::runtime_error{"invalid block_number_or_hash value"};
```

Comment:
Not sure if this case should be handled as exception or by return value containing `std::optional`. In the end of the day it all depends on how this function is supposed to be used and what must be done when `bnoh` is invalid: I guess that an error should be returned, so probably it's fine as it is

Reply:
This solution has been chosen as it seems in line with other `read_*` functions. that return an exception if something went wrong. 
Returning an `std::optional` introduce a little difference...